Core: Move properties

pieterhijma opened the pull request:
This PR introduces functionality to move properties from one object to another.

It includes several things that are clearly separated by different commits.
- small changes to FreeCAD's core to make moving properties possible
- a refactoring of rename property tests to have more consistency between rename and move property tests
- the logic in `App` to allow moving properties
- extensive tests for moving properties
- generalization of the `DlgPropertyLink` dialog to pick objects in general
- a menu item in the property view to move properties

<img width="1675" height="1097" alt="screenshot-move-prop" src="https://github.com/user-attachments/assets/6ab87d09-2e27-4c82-beac-bcff9046df05" />

<!--
The FreeCAD community thanks you for your contribution!
By creating a Pull Request you agree to the contributing policy. The complete policy can be found in the root of the source tree (CONTRIBUTING.md) or at https://github.com/FreeCAD/FreeCAD/blob/main/CONTRIBUTING.md

This template provides guidance on creating a PR that can be reviewed and approved as quickly as possible. Comments may be safely deleted.

Unless you know exactly what you're doing, please leave the checkbox 'Allow edits by maintainers' enabled.  This will allow maintainers to help you.
-->

## Issues
Closes https://github.com/FreeCAD/FreeCAD/issues/22959
Replaces #22960

pieterhijma committed:
Core: Add set defining transaction

This commit allows one to tell the transaction system that it is
currently defining a transaction preventing adding changes.

pieterhijma committed:
Core: Allow global object identifier rename

Remove the assumption that object identifier rename only works on local
changes.

pieterhijma committed:
Core: Refactor rename property tests

pieterhijma committed:
Core: Add move property logic

pieterhijma committed:
Core: Add tests for moving properties

pieterhijma committed:
Gui: Generalize DlgPropertyLink for objects

DlgPropertyLink is now a subclass of DlgDocumentObject for more general
use of selecting objects.

pieterhijma committed:
Gui: Add move property to the property view

kadet1090 commented:
Hi! This PR has seemingly a lot of conflicts due to recent update to the code style. Good news is that in case of that PR formatting changes are the only conflicts so I could update it automatically.

I can update your PR myself or if you want to do it yourself you can find the rebased changes on my fork in branch `pull-23841` (https://github.com/kadet1090/FreeCAD/tree/pull-23841). If you want to use them you can:
1. Check them out locally using `git fetch git@github.com:kadet1090/FreeCAD.git pull-23841`
2. Use `git reset --hard pull-23841` on your PR branch to use my rebased changes
3. `git push --force` to update your PR

If instead you prefer to do the necessary rebasing manually, here is a handy guide: https://github.com/FreeCAD/FreeCAD/discussions/25217.

chennes commented:
Hi @pieterhijma I've flipped this to Draft mode pending resolution of those conflicts. Let us know if you need help with that.

pieterhijma committed:
Gui: Process review comments

Co-authored-by: Max Wilfinger <6246609+maxwxyz@users.noreply.github.com>
Co-authored-by: Chris Hennes <chennes@pioneerlibrarysystem.org>

maxwxyz commented:
@pieterhijma ping for conflicts

maxwxyz commented on src/Gui/Dialogs/DlgDocumentObject.cpp:
```diff
@@ -0,0 +1,998 @@
+
+    connect(ui->buttonBox, &QDialogButtonBox::clicked, this, &DlgDocumentObject::onClicked);
+
+    refreshButton = ui->buttonBox->addButton(tr("Reset"), QDialogButtonBox::ActionRole);
```
```suggestion
    refreshButton = ui->buttonBox->addButton(tr("Refresh"), QDialogButtonBox::ActionRole);
```

Does it really "Reset" or rather "Refresh", as there is also a "Clear" button.

pieterhijma replied:
To be honest, I don't know exactly what those two buttons do. I know that "Reset" is a more appropriate name than "Refresh". To me, both "Clear" and "Reset" clear/reset the selection. 

I think the difference has to do with sublinks but I don't really see the difference. It could be the case that we lost the functionality along the way.

To check on a weekly, create a VarSet and create different types of links, `App::PropertyLink`, `App::PropertyLinkList`, `App::PropertyLinkSub`, `App::PropertyLinkSubList`, etc. Also create a couple of objects, cubes, cylinders, etc. Then in the Property View of the VarSet, give each property a value after which the dialog should open. I couldn't find a noticeable difference between "Clear" and "Reset". This code is essentially moved from DlgPropertyLink.cpp to DlgDocumentObject.cpp and is unchanged.

It may be helpful to try a similar thing on the realthunder branch to see what the supposed functionality is.

Reqrefusion commented:
@pieterhijma This PR will soon be celebrating its first anniversary. Is it ready to be merged?

pieterhijma commented:
> @pieterhijma This PR will soon be celebrating its first anniversary. Is it ready to be merged?

Yep, it's ready for review.

chennes commented on src/App/DocumentObject.cpp:
```diff
@@ -737,6 +737,137 @@ bool DocumentObject::renameDynamicProperty(Property* prop, const char* name)
+    moveExpressionTargetingProp(toBeMovedProp, newProp, targetObj);
+
+    // do not record the the following changes since we are defining a transaction
+    targetObj->_pDoc->setDefiningTransaction(true);
```
This should be RAII'ed -- if either Paste or signalMoveDynamicProperty throws an exception, we're locked in `setDefiningTransaction(true);` forever. Also, it's only doing the `setDefiningTransaction` on the target object's document -- what happens if the expressions that get rewritten below are on the source  document (or some other, third document)?

pieterhijma replied:
You're completely right. I will change it using the RAII pattern.

For your second question, the expressions that are rewritten are not part of transactions anyway. They are silently rewritten: This is the same as with a label change, for example. If an expression refers to a label and the label is changed, the expression is silently rewritten. This example is also verified in the tests both with undo and redo.

chennes commented on src/Gui/propertyeditor/PropertyEditor.cpp:
```diff
@@ -792,6 +793,61 @@ void PropertyEditor::renameProperty(const App::Property& prop)
+            continue;
+        }
+        if (App::Document* doc = propertyDocument(prop->getContainer())) {
+            tid = doc->openTransaction("Move Property");
```
Is there some reason this can only be set for one of the props? Otherwise the `App::GetApplication().commitTransaction(tid);` below is only using the last `tid`, and since `props` is an `unordered_set` there's no way to know which one it's going to be.

pieterhijma replied:
Good point, I wanted to create an issue about this once this was merged. This has been adapted in the rebase to new logic as a result of https://github.com/FreeCAD/FreeCAD/pull/21978. See for example [`removeProperties()`](https://github.com/FreeCAD/FreeCAD/pull/23841/changes#diff-8038059d3ee98b9042f9415c3ce49663ba1277b9fcd1305942fdd7264ba0f06dL946) that is not part of this PR . 

I assume that the various transaction ids are linked and that the last transaction id commits all transactions. @theo-vt, can you confirm this?

chennes commented on src/App/Document.h:
```diff
@@ -1053,6 +1053,17 @@ class AppExport Document: public PropertyContainer
+     * @brief Register in a transaction that a property move has been arranged.
+     *
+     * @param[in] obj The object whose property is moved.
+     * @param[in] obj The property that is moved.
```
```suggestion
     * @param[in] obj The object whose property is moved.
     * @param[in] toBeMovedProp The property that is moved.
```

chennes commented on src/App/DocumentObject.cpp:
```diff
@@ -737,6 +737,137 @@ bool DocumentObject::renameDynamicProperty(Property* prop, const char* name)
+    // Phase 2: Move an expression that targets the current property
+    moveExpressionTargetingProp(toBeMovedProp, newProp, targetObj);
+
+    // do not record the the following changes since we are defining a transaction
```
```suggestion
    // do not record the following changes since we are defining a transaction
```

chennes reviewed (commented):
In addition to the comments below, if I select "move property" and then just click "OK" in the resulting selection dialog (without selecting anything), FreeCAD crashes. Just needs a check on the returned selection before use.

pieterhijma committed:
Core: Add a ScopeGuard for marking transactions

This commit adds a new ScopeGuard object that can be easily reused for
settings that need to be reset after a scope ends.  This is used to mark
that a transaction is being defined in a document.

pieterhijma commented:
> In addition to the comments below, if I select "move property" and then just click "OK" in the resulting selection dialog (without selecting anything), FreeCAD crashes. Just needs a check on the returned selection before use.

Thanks, I've already pushed some changes, but I still need to look into the dialog crash.

pieterhijma commented:
> > In addition to the comments below, if I select "move property" and then just click "OK" in the resulting selection dialog (without selecting anything), FreeCAD crashes. Just needs a check on the returned selection before use.

Indeed, I added a check for an empty selection.

It's ready for review again.

chennes commented:
OK, I poked a bit at the transaction stuff here: see the last commit I pushed your way (you can drop it if you don't want to include it, it was really sort of exploratory). At any rate, the "test" simply replicates what your current code is doing, and fails its assertions at the end because it *does not* restore both properties.

(Side note: to really write the test properly, `moveProperties` would need to be exposed in such a way that the test code could call it. Actually that wouldn't be a terrible refactor to do anyway, but I am not "demanding" it here.)

pieterhijma commented:
> OK, I poked a bit at the transaction stuff here: see the last commit I pushed your way (you can drop it if you don't want to include it, it was really sort of exploratory). At any rate, the "test" simply replicates what your current code is doing, and fails its assertions at the end because it _does not_ restore both properties.

This is not really what I meant. @theo-vt knows the details better and it would be good if he could comment. 

I'm pretty sure about the following: Before the multi file editing PR, you could do something like you are trying to do in the test: In the property view, you would be able to select multiple properties and delete them in one go. You would get only one transaction and an undo would restore all those properties at once. This is also how I initially implemented move properties. I even changed the transaction name from "Move Property" for one property to "Move Properties" for multiple properties.

The following I'm not so sure about, but it is what I infer from the code: Now that the multi-file editing PR has been merged, the logic has changed. I noticed that remove properties does not support a single transaction for removing multiple properties any longer (or maybe it does, but the code has changed in that respect). Although there is only one commit on the last ID, under the hood, multiple transactions are committed and multiple entries are added to the undo list. I think this is a conscious decision by @theo-vt, probably to make his work a bit easier, while still have some sort of overarching transaction ID, but it would be good if @theo-vt can confirm this.

While rebasing this PR, I've adopted it to this approach for consistency, the one of remove properties. This means creating multiple transactions and doing a commit on the last ID, which commits all the previous transactions as well. This practically means that if you go in the Property View, you select *n* properties and delete those, you will get *n* undo items. The same should now be true for move properties. 

This means that your added test is half-way between the old situation and the new situation. To make your test work, it should be this:

```c++
// Moving several properties in one action must form a single undo step. This mirrors the loop in
// the property editor that opens a transaction per property and commits only the last id.
TEST_F(MoveProperty, moveMultipleIsAtomic)
{
    // Arrange
    doc1->setUndoMode(1);
    auto* prop2 = freecad_cast<App::PropertyInteger*>(
        varSet1Doc1->addDynamicProperty("App::PropertyInteger", "Variable2", "Variables")
    );
    prop2->setValue(value + 1);
    std::vector<App::Property*> props {prop, prop2};

    // Act
    int tid = 0;
    for (auto* property : props) {
        tid = doc1->openTransaction("Move Property");
        varSet1Doc1->moveDynamicProperty(property, varSet2Doc1);
    }
    App::GetApplication().commitTransaction(tid);

    // Assert: a single move action produces a single undo step
    EXPECT_EQ(doc1->getAvailableUndos(), 2);

    // Assert: one undo restores both properties to the source container
    EXPECT_TRUE(doc1->undo());
    EXPECT_TRUE(doc1->undo());
    EXPECT_NE(varSet1Doc1->getDynamicPropertyByName("Variable"), nullptr);
    EXPECT_NE(varSet1Doc1->getDynamicPropertyByName("Variable2"), nullptr);
    EXPECT_EQ(varSet2Doc1->getDynamicPropertyByName("Variable"), nullptr);
    EXPECT_EQ(varSet2Doc1->getDynamicPropertyByName("Variable2"), nullptr);
}
```

For comparison, this test should work for removal of properties:

```c++
// Moving several properties in one action must form a single undo step. This mirrors the loop in
// the property editor that opens a transaction per property and commits only the last id.
TEST_F(MoveProperty, removeMultipleIsAtomic)
{
    // Arrange
    doc1->setUndoMode(1);
    auto* prop2 = freecad_cast<App::PropertyInteger*>(
        varSet1Doc1->addDynamicProperty("App::PropertyInteger", "Variable2", "Variables")
    );
    prop2->setValue(value + 1);
    std::vector<App::Property*> props {prop, prop2};

    // Act
    int tid = 0;
    for (auto* property : props) {
        tid = doc1->openTransaction("Remove Property");
        varSet1Doc1->removeDynamicProperty(property->getName());
    }
    App::GetApplication().commitTransaction(tid);

    // Assert: a single move action produces a single undo step
    EXPECT_EQ(doc1->getAvailableUndos(), 2);

    // Assert: one undo restores both properties to the source container
    EXPECT_TRUE(doc1->undo());
    EXPECT_TRUE(doc1->undo());
    EXPECT_NE(varSet1Doc1->getDynamicPropertyByName("Variable"), nullptr);
    EXPECT_NE(varSet1Doc1->getDynamicPropertyByName("Variable2"), nullptr);
}
```

The difference is that there are 2 undo items available and undo is execute twice.

I hope this clears things up!

maxwxyz commented:
@pieterhijma CI fails

pieterhijma commented:
> @pieterhijma CI fails

Yes I know, that is caused by a couple of tests that @chennes added that I think are incorrect. I'll ask from a re-review from @chennes.

chennes commented:
OK, I re-reviewed and understand the intent. I think the code itself *looks* like it's constructing a single transaction, so is a bit confusing, but I think it's OK. I've dropped the test I added. It looks like maybe the existing tests need a conflict resolution before merging though.

pieterhijma commented:
I resolved the merge conflict.
